### PR TITLE
Prioritized URLRequest signature, organized overloads

### PR DESCRIPTION
**Pull request by bdrelling**
**Description**
The intent of this PR is to change the primary `request` and `requestPublisher` signatures for `HTTPClient` to prefer `URLRequest` instead of any custom object for this SDK or a collection of properties to build a request. This allows us to pass in existing `URLRequest` objects without issue. 

In addition, I added a convenience initializer to `URLRequest` and cleaned up a bit of the shared code as well. Hopefully, this makes the primary `HTTPClient` and `HTTPClient+Combine` classes more maintainable overall.

### CODE REVIEW

File: Sources/UtilityBeltNetworking/Extensions/DataResponse+Failure.swift [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// Copyright © 2021 SpotHero, Inc. All rights reserved.
+
+extension DataResponse {
```
Comment (bdrelling) [inline Sources/UtilityBeltNetworking/Extensions/DataResponse+Failure.swift:3]: This extension was copied from `HTTPClient` and moved into its own file.

File: Sources/UtilityBeltNetworking/Extensions/URLRequest+Init.swift [hunk 2]
```diff
@@ -0,0 +1,30 @@
+// Copyright © 2021 SpotHero, Inc. All rights reserved.
+
+import Foundation
+
+public extension URLRequest {
```
Comment (bdrelling) [inline Sources/UtilityBeltNetworking/Extensions/URLRequest+Init.swift:5]: This extension was based on `HTTPClient.configuredURLRequest`.

File: Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestOverloads.swift [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        dispatchQueue: DispatchQueue = .main,
+        decoder: JSONDecoder = JSONDecoder(),
+        completion: DecodableTaskCompletion<T>? = nil
+    ) -> Request? {
+        var convertedRequest: URLRequest
+        
+        do {
+            convertedRequest = try request.asURLRequest()
+        } catch {
+            DispatchQueue.main.async {
```
Comment (khaptonstall) [inline Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestOverloads.swift:153]:
```suggestion
            dispatchQueue.async {
```

File: Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+    @available(iOS 13, macOS 10.15, tvOS 13.0, watchOS 6, *)
+    public extension HTTPClient {
+        // MARK: Data Response
+    
+        /// Creates a request publisher which fetches raw data from an endpoint.
+        /// - Parameter request: An object defining properties to make the request with.
+        /// - Parameter validators: An array of validators that will be applied to the response.
+        /// - Parameter interceptor: An object that can intercept the url request. Defaults to `nil`.
+        /// - Parameter dispatchQueue: The `DispatchQueue` to make the request on.
```
Comment (khaptonstall) [inline Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift:15]:
```suggestion
        /// - Parameter dispatchQueue: The `DispatchQueue` to make the request on. Defaults to `.main`.
```

File: Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        }
+    
+        // MARK: Decodable Object Response
+    
+        /// Creates a request publisher which fetches raw data from an endpoint and decodes it.
+        /// - Parameter request: The `URLRequest` to make the request with.
+        /// - Parameter validators: An array of validators that will be applied to the response. Defaults to ensuring a JSON mime type.
+        /// - Parameter interceptor: An object that can intercept the url request. Defaults to `nil`.
+        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data.
+        /// - Parameter dispatchQueue: The `DispatchQueue` to make the request on.
```
Comment (khaptonstall) [inline Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift:123]:
```suggestion
        /// - Parameter dispatchQueue: The `DispatchQueue` to make the request on. Defaults to `.main`.
```

File: Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+            )
+        }
+    
+        // MARK: Decodable Object Response
+    
+        /// Creates a request publisher which fetches raw data from an endpoint and decodes it.
+        /// - Parameter request: The `URLRequest` to make the request with.
+        /// - Parameter validators: An array of validators that will be applied to the response. Defaults to ensuring a JSON mime type.
+        /// - Parameter interceptor: An object that can intercept the url request. Defaults to `nil`.
+        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data.
```
Comment (khaptonstall) [inline Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift:122]: This should come after the `dispatchQueue` to match the method signature.

Comment (khaptonstall) [inline Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift:122]:
```suggestion
        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data. Defaults to `JSONDecoder()`.
```

File: Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        /// Creates a request publisher which fetches raw data from an endpoint and decodes it.
+        /// - Parameter url: The URL for the request. Accepts a URL or a String.
+        /// - Parameter method: The HTTP method for the request. Defaults to `GET`.
+        /// - Parameter parameters: The parameters to be converted into a String-keyed dictionary to send in the query string or HTTP body.
+        /// - Parameter headers: The HTTP headers to send with the request.
+        /// - Parameter encoding: The parameter encoding method. If nil, uses the default encoding for the provided HTTP method.
+        /// - Parameter validators: An array of validators that will be applied to the response. Defaults to ensuring a JSON mime type.
+        /// - Parameter interceptor: An object that can intercept the url request. Defaults to `nil`.
+        /// - Parameter dispatchQueue: The dispatch queue on which the response will be published. Defaults to `.main`.
+        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data.
```
Comment (khaptonstall) [inline Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift:158]:
```suggestion
        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data. Defaults to `JSONDecoder()`.
```

File: Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        /// Creates a request publisher which fetches raw data from an endpoint and decodes it.
+        /// - Parameter url: The URL for the request. Accepts a URL or a String.
+        /// - Parameter method: The HTTP method for the request. Defaults to `GET`.
+        /// - Parameter parameters: The `Encodable` object to be converted into a String-keyed dictionary to send in the query string or HTTP body.
+        /// - Parameter headers: The HTTP headers to send with the request.
+        /// - Parameter encoding: The parameter encoding method. If nil, uses the default encoding for the provided HTTP method.
+        /// - Parameter validators: An array of validators that will be applied to the response. Defaults to ensuring a JSON mime type.
+        /// - Parameter interceptor: An object that can intercept the url request. Defaults to `nil`.
+        /// - Parameter dispatchQueue: The dispatch queue on which the response will be published. Defaults to `.main`.
+        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data.
```
Comment (khaptonstall) [inline Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift:205]:
```suggestion
        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data. Defaults to `JSONDecoder()`.
```

**Comment by bdrelling** [comment]
@khaptonstall Thanks for the thorough review, updated all the outdated comments I could find across those four files.

Regarding the test failures, this was because of the `HTTPClientTests` file, which was primarily handling timeouts. Previously, it would use the `HTTPClient` method of configuring a `URLRequest`, but given that this method was removed in favor of a direct initializer on `URLRequest`, I updated the test cases to work with the `URLRequest` object and a new set of extensions directly.

**Review by khaptonstall — APPROVED** [review]

**Pull request merged**